Stratis.Bitcoin/Base/ChainState.cs

```diff
@@ -1,78 +1,81 @@
 using NBitcoin;
-using System;
-using System.Collections.Generic;
-using System.Linq;
-using System.Threading;
 using Stratis.Bitcoin.Interfaces;
+using System;
+using System.Collections.Concurrent;
 
 namespace Stratis.Bitcoin.Base
 {
     public class ChainState
     {
-        private readonly FullNode fullNode;
+        /// <summary>The fullnode interface.</summary>
+        private readonly IFullNode fullNode;
+
+        /// <summary>The last time the <see cref="ibdLastResult"/> was updated.</summary>
+        private long ibdLastUpdate;
+        
+        /// <summary>A cached result of the IBD method.</summary>
+        private bool ibdLastResult;
 
-        private long lastUpdate;
-        private bool lastResult;
+        /// <summary>A collection of blocks that have been found to be invalid.</summary>
+        internal ConcurrentDictionary<uint256, uint256> InvalidBlocks;
```

Comment:
I think this should not be internal, but private. The only place we use it is in `CheckAnnouncedBlocks` and that can easily be rewritten in a way that this can be private. All we need to do is to expose number of elements here and then to avoid accessing ContainsKey directly and just use IsMarkedInvalid which we already have.

Reply:
OK, I correct myself. Just delete CheckAnnouncedBlocks completely, it is not used anywhere and it is flawed - it uses EnumerateToGenesis, which seems crazy. But not used anywhere, so just delete it and make this private.

Reply:
erm, now I realized you are using concurrent dictionary instead of HashSet ... the correct thing you would like to use here is ConcurrentHashSet which does not exist, so you can either implement it, or just use a lock

however, if you still insist on using concurrent dictionary instead (which is wrong ;) ) then please, at least make it CD<uint256, byte> not to occupy memory with unused 32 bytes of a thing

Reply:
But the value in the dictionary doe not get cloned so its just a reference (no extra memory is allocated)
There is a ConcurrenctHashset nuget package heavily used we did not include it the project (yet).

Yeah that method `CheckAnnouncedBlocks` is anyway not used but I keep it there as it may be used when we implement banning.

Reply:
hmm you are correct with the reference, but still are we not talking 1 reference amount of memory vs 1 byte? anyway, I would just go for hashset with lock, that's what you want to use, you don't want dictionary ...

that CAB function is flawed entirely, we should remove it for sure and make this private and only when we need something then to implement it correctly

Reply:
Ok fair enough but I will leave this of another PR as I am already doing too much out of this task.

Reply:
agree

Reply:
this is done